File: src/daytrace/diagnostics.py
from __future__ import annotations

from collections import Counter
from collections.abc import Iterable

from daytrace.models import DiagnosticCode, DiagnosticCount


_MESSAGE = {
    DiagnosticCode.NON_POSITIVE_EVENT: "non-positive ActivityWatch event",
    DiagnosticCode.NAIVE_TIMESTAMP: "event with a naive timestamp",
    DiagnosticCode.UNSUPPORTED_BUCKET: "unsupported ActivityWatch bucket",
    DiagnosticCode.WINDOW_CONFLICT: "conflicting foreground interval",
    DiagnosticCode.SANITIZED_FIELD: "captured field containing private data",
}


class DiagnosticCollector:
    def __init__(self) -> None:
        self._counts: Counter[DiagnosticCode] = Counter()

    def add(self, code: DiagnosticCode, count: int = 1) -> None:
        if count > 0:
            self._counts[code] += count

    def snapshot(self) -> tuple[DiagnosticCount, ...]:
        return tuple(
            DiagnosticCount(code, self._counts[code])
            for code in DiagnosticCode
            if self._counts[code]
        )


def diagnostic_messages(counts: Iterable[DiagnosticCount]) -> tuple[str, ...]:
    return tuple(
        f"ignored {item.count} {_MESSAGE[item.code]}"
        + ("" if item.count == 1 else "s")
        for item in counts
    )

File: packages/daytrace-core/src/diagnostics.ts
import type { DiagnosticCode, DiagnosticCount } from "./models.js";

const ORDER: readonly DiagnosticCode[] = [
  "non-positive-event",
  "naive-timestamp",
  "unsupported-bucket",
  "window-conflict",
  "sanitized-field",
];

const MESSAGES: Readonly<Record<DiagnosticCode, string>> = {
  "non-positive-event": "non-positive ActivityWatch event",
  "naive-timestamp": "event with a naive timestamp",
  "unsupported-bucket": "unsupported ActivityWatch bucket",
  "window-conflict": "conflicting foreground interval",
  "sanitized-field": "captured field containing private data",
};

export class DiagnosticCollector {
  readonly #counts = new Map<DiagnosticCode, number>();

  add(code: DiagnosticCode, count = 1): void {
    if (count > 0) this.#counts.set(code, (this.#counts.get(code) ?? 0) + count);
  }

  snapshot(): readonly DiagnosticCount[] {
    return ORDER.flatMap((code) => {
      const count = this.#counts.get(code) ?? 0;
      return count > 0 ? [{ code, count }] : [];
    });
  }
}

export function diagnosticMessages(counts: readonly DiagnosticCount[]): readonly string[] {
  return counts.map(({ code, count }) => (
    `ignored ${count} ${MESSAGES[code]}${count === 1 ? "" : "s"}`
  ));
}
